### ISSUE TITLE
Upgrade POA bridge to 2.x bridge contracts

### ISSUE THREAD

**Issue by akolotov** [poanetwork/poa-bridge-contracts#137]
Under this issue it is required to perform a research and prepare a process to upgrade the POA bridge to use 2.x contracts.

**Comment by patitonar** [comment]
I'm adding this comment to start the discussion about the possible approach to migrate POA bridge contracts. I added information about v1.x and v2.x contracts, some consideration and possible steps for migration. 

## Current deployed POA bridge
- v1.x contracts
- POA rust bridge
- Bridge UI for v1.x contracts
- Bridge monitor for v1.x contracts


## We need to upgrade to
- v2.x contracts
- token-bridge
- Bridge UI for v2.x contracts
- Bridge monitor for v2.x contracts

## Contract differences
Differences on events emitted and methods called by validators

| Version | V1 | V2 | 
| ------------- | ------------- | ------------- |
**Events** |  | 
Home | Deposit | UserRequestForSignature
Home |  | CollectedSignatures
Foreign | Withdraw | UserRequestForAffirmation
Foreign | CollectedSignatures |
**Methods** |  | 
Home | withdraw | submitSignature
Home |  | executeAffirmation
Foreign | deposit | executeSignatures
Foreign | submitSignature | 


## Considerations
1. A transfer could be started with v1 contracts and not finished processing by validators before upgrade of contracts takes place. That's why the new implementation of contract should allow validators to finish processing previous transfers.

## Steps for migration
Here are some possible steps for contracts migration:
1. Prepare and deploy a contract version based on v2 contracts with support for methods used by validators on v1 in order to be able to finish processing any pending transfer. 
    - Methods to support on Home for validators:
        - `withdraw` (and other methods it uses)
            - MessageV1 library
            - withinForeignLimit
            - withdraws
            - setWithdraws
    - Methods to support on Foreign for validators:
        - `submitSignature` (and other methods it uses)
            - MessageV1 library
        - `deposit` (and other methods it uses)
            - withinHomeLimit
            - depositsSigned
            - setDepositsSigned
            - numDepositsSigned
            - setNumDepositsSigned
2. Start token-bridge for validators
3. Upgrade home bridge impl
    - add method to update params to be called by `upgradeToAndCall`:
        - set dailyLimit = homeDailyLimit 
        - set executionDailyLimit = foreignDailyLimit 
        - set executionMaxPerTx = foreignMaxPerTx 
4. Upgrade foreign bridge impl
    - add method to update params to be called by `upgradeToAndCall`:
        - set dailyLimit = foreignDailyLimit 
        - set executionDailyLimit = homeDailyLimit 
        - set executionMaxPerTx = homeMaxPerTx 
5. Deploy Bridge-ui for v2 contracts
6. Once there is no pending Tx from v1 events. Stop bridge-rust validators.
7. Deploy Bridge-monitor for v2 contracts
8. Upgrade Home bridge impl to v2.x contracts (without methods for v1 events)
9. Upgrade Foreign bridge impl to v2.x contracts (without methods for v1 events)

I think steps `8` and `9` could be optional, methods for v1 events could just be part of contracts and be removed next time an upgrade is needed on the contracts.

### Questions

1. Since both app will be sending transactions from same validator's addresses on a given period of time, the possibility of sending transactions with the same nonce could happen. I'm not familiar on how Rust implementation works, but on token-bridge version if an error by nonce is presented, it will read the current nonce and try again, so I don't think there will be an issue from this side.

**Comment by akolotov** [comment]
Thanks for analysis.

First of all, we can simply stop all bridge operation by setting the daily limit in 0 on both side and waiting some amount of time. It should reduce number of work to change V2 version of contracts.

Secondly, I am worrying about storage and already relayed transactions. If we just apply new implementations will it be possible to replay previous transactions (apply them once again) or protections in the new implementation will work with old data from the storage?

Another question for other configuration parameters. Will it be enough to update limits as you stated above? Are any new configuration parameters appeared in V2? Could you compare this also?

**Comment by patitonar** [comment]
> First of all, we can simply stop all bridge operation by setting the daily limit in 0 on both side and waiting some amount of time. It should reduce number of work to change V2 version of contracts.

Right, if stopping operations and waiting to all transfer to finish is an option then we can upgrade to V2 directly without the need of supporting V1 methods.

> Secondly, I am worrying about storage and already relayed transactions. If we just apply new implementations will it be possible to replay previous transactions (apply them once again) or protections in the new implementation will work with old data from the storage?

The only method that I think is exposed to users to try to replay previous transactions is `executeSignatures`. The user could get the message and signatures of validators for an already relayed message and try to execute them again.

On V1, message length was `116` composed by:
- recipientAddress
- value
- transactionHash
- gasPrice

On V2, message length is `104` composed by:
- recipientAddress
- value
- transactionHash
- bridgeAddress

On `executeSignatures` on calls to Message library: `Message.hasEnoughValidSignatures` and `Message.parseMessage` it checks that the message length must be `104`. Also, there is a check for bridgeAddress from the message: `require(contractAddress == address(this));`.
So trying to execute signatures with a message from v1 will fail.

> Another question for other configuration parameters. Will it be enough to update limits as you stated above? Are any new configuration parameters appeared in V2? Could you compare this also?

The only differences I found on parameters from V1 to V2 are the ones related to execution limit as detailed on the previous comment.
- On Home:
    - V1:
         - homeDailyLimit
         - foreignDailyLimit
         - foreignMaxPerTx
    - V2:
         - dailyLimit
         - executionDailyLimit
         - executionMaxPerTx
- On Foreign:
    - V1:
         - foreignDailyLimit
         - homeDailyLimit
         - homeMaxPerTx
    - V2:
         - dailyLimit
         - executionDailyLimit
         - executionMaxPerTx

**Comment by akolotov** [comment]
OK. Great! So, the next step is to try to test the upgrade: setup the V1 bridge between Ropsten and Sokol with usage of the same contracts (taken from your gist or from the validated code in Etherescan and Blockscout), make ten or twenty transactions and upgrade it with the latest V2 contracts, make another dozen of transactions. Then rollback block numbers in the watchers to the blocks where the V1 contracts were deployed and restart the watchers. Look carefully what will happen.

**Comment by patitonar** [comment]
I tested the upgrade, the bridge operations worked correctly. Here are the details and steps:

Home Network: `Sokol`
Foreign Network: `Ropsten`

1. Started with V1 bridge contract (used implementation from gist), rust bridge, bridge monitor and bridge UI v1.
2. Send some transactions.
3. Based on v2.2.0 contracts added a method to call on the upgrade:

On  `HomeBridgeNativeToErc` ([Gist](https://gist.github.com/patitonar/4cd31558aa3a46ac07d8f0662161c6f7)):
```solidity
    function upgradeToV220() public {
        require(!boolStorage[keccak256(abi.encodePacked("isUpgradedToV220"))]);
        uintStorage[keccak256(abi.encodePacked("dailyLimit"))] = uintStorage[keccak256("homeDailyLimit")];
        uintStorage[keccak256(abi.encodePacked("executionDailyLimit"))] = uintStorage[keccak256("foreignDailyLimit")];
        uintStorage[keccak256(abi.encodePacked("executionMaxPerTx"))] = uintStorage[keccak256("foreignMaxPerTx")];
        boolStorage[keccak256("isUpgradedToV220")] = true;
    }
```

On ` ForeignBridgeNativeToErc` ([Gist](https://gist.github.com/patitonar/20e414002619728e835d5ce4421b46e7)):

```solidity
    function upgradeToV220() public {
        require(!boolStorage[keccak256(abi.encodePacked("isUpgradedToV220"))]);
        uintStorage[keccak256(abi.encodePacked("dailyLimit"))] = uintStorage[keccak256("foreignDailyLimit")];
        uintStorage[keccak256(abi.encodePacked("executionDailyLimit"))] = uintStorage[keccak256("homeDailyLimit")];
        uintStorage[keccak256(abi.encodePacked("executionMaxPerTx"))] = uintStorage[keccak256("homeMaxPerTx")];
        boolStorage[keccak256("isUpgradedToV220")] = true;
    }
```

4. Started token-bridge
5. Deployed new implementation contracts and upgraded the contracts using `upgradeToAndCall`
6. Started Bridge UI V2
7. Started Bridge monitor V2
8. Sent transactions using Bridge UI. Token-bridge worked correctly. Rust bridge was also running but didn't receive any event. Every transfer worked correctly.
9. I reset the last processed block on all watchers to `0`. After processing all blocks again, watchers only received events emitted after the upgrade to V2 (which is Ok because of the name of events) and discarded them because they were already processed. Bridge operations continued normally.

After step `6`. While using Bridge UI, despite of transfer working correctly, I found 2 issues on the information displayed on `status` and `statistics` pages.
- On `status` page bridge validator list is not displayed. It listen to `ValidatorAdded` and  `ValidatorRemoved` events to build the list. I found that there is an error while trying to decode the parameter of the event and it is related to a change on how the event is defined on V2 and V1 bridge validators contracts. On V2 the address parameter is indexed while on V1 it is not. So bridge UI is using V2 ABI to get the events which throws the error. **This same issue affects bridge monitor** causing to crash while getting the validators. For this issue I can add a fix to try to use V1 ABI if using V2 fails on bridge ui and bridge monitor.

- On `statistics` page: On the count of number of deposits and withdraw it only display the number of transactions performed after the upgrade to V2. This is because to calculate the numbers it listen to the new events name, so previous transactions that generated `Deposit` and `Withdraw` events are not count. For this issue I can add a fix to also listen to these events using V1 ABI too, so we can get all the total transaction count.
